### PR TITLE
Reinforcement Learning using DyNet: DQN and DDPG

### PR DESCRIPTION
**Pull request by zhisbug**
https://github.com/clab/dynet/issues/1290

### CODE REVIEW

File: examples/reinforcement-learning/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+* [OpenAI Gym](https://github.com/openai/gym)
+
+##  Classical Control with Discrete Action Space using DQN
+### Run training
+```
+python main_dqn.py --env_id 0
+```
+where `env_id` chooses one of the classic control environment from OpenAI Gym, which includes [CartPole](https://gym.openai.com/envs/CartPole-v1/), [MountainCar](https://gym.openai.com/envs/MountainCar-v0/), and [Acrobot](https://gym.openai.com/envs/Acrobot-v1/), all with **discrete action space**.
+
+To enable Double Q-Learning, Prioritized Replay and Dueling Network Arhictecture, we can add the following tags, respectively:
```
Comment (neubig) [inline examples/reinforcement-learning/README.md]: Arhictecture -> Architecture

File: examples/reinforcement-learning/ddpg.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import dynet as dy
+import numpy as np
+
+from memory import Memory
+from network import MLP
+
+
+class DDPG:
```
Comment (neubig) [inline examples/reinforcement-learning/ddpg.py]: Could you add a comment with a reference that readers can reference to understand this method?

File: examples/reinforcement-learning/dqn.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+import dynet as dy
+import numpy as np
+
+from memory import Memory, PrioritizedMemory
+
+
+class DeepQNetwork(object):
```
Comment (neubig) [inline examples/reinforcement-learning/dqn.py]: Could you add a comment with a reference that readers can reference to understand this method?

File: examples/reinforcement-learning/memory.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import numpy as np
+from math import log, ceil
+
+from reduce_tree import ReduceTree, SumTree
+
+
+class Memory(object):
```
Comment (neubig) [inline examples/reinforcement-learning/memory.py]: Could you add comments?

File: examples/reinforcement-learning/memory.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    def store(self, exp):
+        self.memory[self.idx % self.size] = exp
+        self.idx += 1
+
+    def sample(self, batch_size):
+        indices = np.random.choice(min(self.idx, self.size), batch_size)
+        return self.memory[indices]
+
+
+class PrioritizedMemory(object):
```
Comment (neubig) [inline examples/reinforcement-learning/memory.py]: Could you add comments?

**Comment by zhisbug** [comment]
@neubig : @zhiyong1997 addressed your comment. Could you take a look at why the CI failed? It seems to be a CI environment issue to me.

**Comment by pmichel31415** [comment]
Yes I'm not sure if the travis test failing are a transient error or if they're due to this pull request. @danielhers  I think you were the one to set up travis CI, is there a way you can relaunch the tests. Even better, can you add me (and @neubig ) to the project on travis so that we can also rerun tests? Thanks!

**Comment by danielhers** [comment]
I relaunched the build. Please create an account on AppVeyor and send me your email address so I'll add you to the collaborators.

**Comment by zhiyong1997** [comment]
Thanks for rejudging the PR ! I have now created an account on AppVeyor
with this email: zhiyong.fang.1997@gmail.com.

*Zhiyong Fang*

Acm Hornored Class, Zhiyuan College, Shanghai Jiao Tong University
Mobile : +86-133-9126-7072
Email  : zhiyong.fang.1997@gmail.com

**Comment by danielhers** [comment]
Sorry, I meant @pmichel31415  - should I add @zhiyong1997 too?

**Comment by zhiyong1997** [comment]
Oh. Just leave it. I'm actually not a collaborator.
*Zhiyong Fang*

Acm Hornored Class, Zhiyuan College, Shanghai Jiao Tong University
Mobile : +86-133-9126-7072
Email  : zhiyong.fang.1997@gmail.com

**Comment by pmichel31415** [comment]
@danielhers my email is pmichel31415@gmail.com

Thanks!

**Comment by danielhers** [comment]
@pmichel31415 I think you should be able to (re-)run builds now. Anyway, this PR passed (it was indeed a transient fail).

**Comment by pmichel31415** [comment]
Cool thank you!

**Comment by neubig** [comment]
I think it's better to merge this than wait...

**Pull request merged**